### ISSUE TITLE
Handle metadata imprecisions: Single pixel overlaps

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-tasks-core#91]
While running the example in `04_cardio_multiwell`, from the `fractal` repo, the file `/data/active/fractal/tests/20220920_10well_5x5/20200812-CardiomyocyteDifferentiation14-Cycle1.zarr/F/09/0/0/.zarray` reads
```
{
    "chunks": [
        1,
        1,
        2160,
        2560
    ],
    "compressor": {
        "blocksize": 0,
        "clevel": 5,
        "cname": "lz4",
        "id": "blosc",
        "shuffle": 1
    },
    "dimension_separator": "/",
    "dtype": "<u2",
    "fill_value": 0,
    "filters": null,
    "order": "C",
    "shape": [
        3,
        19,
        10800,
        12799
    ],
    "zarr_format": 2
}
```

Note the 12799 dimension along the X direction.

**Comment by tcompa** [comment]
We should at least include a check like
```python
if shape_x % img_size_x > 0:
    raise Exception(...)
```

**Comment by jluethi** [comment]
Is this related to the precision of the pixel sizes? Or any other idea where the off-by-one error is coming from?

**Comment by tcompa** [comment]
For the record, this the `.zarray` for a simpler example (a single well with 9x8 FOVs, in `/data/active/fractal/3D/PelkmansLab/CardiacMultiplexing/Cycle1_9x8_singleWell`), which may be simpler to debug:
```
{
    "chunks": [
        1,
        1,
        2160,
        2560
    ],
    "compressor": {
        "blocksize": 0,
        "clevel": 5,
        "cname": "lz4",
        "id": "blosc",
        "shuffle": 1
    },
    "dimension_separator": "/",
    "dtype": "<u2",
    "fill_value": 0,
    "filters": null,
    "order": "C",
    "shape": [
        3,
        19,
        19439,
        20479
    ],
    "zarr_format": 2
```

**Comment by tcompa** [comment]
The `get_ROIs_bounding_box` function was returning an array of floats for each shape element, instead of a single int. I don't know yet whether this explains the wrong shape, but for sure it had to be fixed.

**Comment by tcompa** [comment]
It looks like the error is in the metadata file.

An image should have a X size of 416 micrometer (corresponding to 2560 pixels, with 0.1625 micrometer/pixel).
If I look at `data/active/fractal/3D/PelkmansLab/CardiacMultiplexing/Cycle1_9x8_singleWell/MeasurementData.mlf`, I see that the min/max values of X are:
```
X_min=-1448.3   # e.g. in 20200812-CardiomyocyteDifferentiation14-Cycle1_B03_T0002F009L01A01Z01C01.tif
X_max=1463.6    # e.g. in 20200812-CardiomyocyteDifferentiation14-Cycle1_B03_T0002F008L01A01Z01C01.tif
```
The difference between these two values is 2911.9 micrometers (instead of the expected 7*416=2912 micrometers), which is not an integer multiple of 416.

@jluethi @gusqgm, can you let us know how this works and how to proceed?
It's not clear whether this kind of error is a surprising behavior or something that we should support.

**Comment by jluethi** [comment]
Thanks for tracing it there, I'll have a look into this. It's a bit surprising to me, but maybe those positions get rounded and we'll need to find a way to work with that precision. I'll look into it a bit more to get an idea. If there really is rounding of those positions, we'd have to check if there are potential off-by-1 errors also in the middle of a well, leading to a column of zeros or a column that gets overwritten by 2 FOVs due to that rounding.

**Comment by jluethi** [comment]
Big picture: We'll need to actually handle that. It's not that the metadata is wrong, just that the precision is just slightly too low for how we're trying to use it here. We will need to find a workaround for that though.

Some first thoughts into this, but not finished yet:
The precision of positional metadata of the yokogawa is limited to 0.1 um. That is somewhat reasonable given the precision of the actual stage.
Pixel sizes in our test set are 0.1625 um. Other typical pixel sizes are 0.325 and 0.108333333. Thus, the precision of how stage positions are saved is unfortunately very close to the step size of our pixels, making an off-by-one error something we need to be careful about.

The lack of precision is not due to our rounding. e.g. in the case where the calculation says it's a distance of 2911.9 micrometers. With a pixel size of 0.1625, that would be 17’919.384 pixels. But we'd actually expect 17920 pixels. Thus, even rounding would get us to an off-by-one error in these calculations vs. the underlying assumed positions.

A few things worth looking into next:
1. If we just do those calculations for each FOV separately & round, would we be save then (against both + & - 1 errors)?
2. Do we have a way to calculate whether we would have ROI overlap and handle the case of an (unexpected) single-pixel overlap? Do the individual ROIs actually have a single pixel overlap or would we have an off-by-one error when trying to access the array somewhere?

**Comment by jluethi** [comment]
This is trickier than I expected. The same error would pop up if we just go FOV-by-FOV.

Here are the FOV x_coordinate positions for the first row in a 9x8 well (23 well dataset, B03)
<img width="276" alt="Bildschirmfoto 2022-09-23 um 16 45 43" src="https://user-images.githubusercontent.com/18033446/191987962-e00b724a-2ac1-407d-9fba-d7d3651840f4.png">

And here are the distances between the all FOVs based on the x_micrometer positions, as well as the corresponding pixel shifts when divided by the pixel size

<img width="708" alt="Bildschirmfoto 2022-09-23 um 16 49 21" src="https://user-images.githubusercontent.com/18033446/191988835-59d2c0e9-2d4a-4da0-9cf7-1b553318768e.png">

One interesting observation here: The rounding error always occurs in the same place. Independent of the well or the row one looks at, it's always FOV 3 that has a distance of 415.9um, when it should really have a distance of 416 um.

I think this shows that it's really a rounding happening in the way Yokogawa saves this metadata, robustly rounding their stage position metadata.

-----

### What do we do with this?

On one level, we want to parse the real metadata and it's a tricky discussion on whether we should change any of the actual metadata when parsing data (preferably, we'd just save the actual metadata the microscope gave us).
Unfortunately, in this case, the microscope metadata is not accurate enough for what we need, making consistent tiny rounding errors that can shift things by a single pixel.

I think we could detect this after having parsed the table and correct the table in this simplest case (i.e. check that all FOVs don't have a single pixel overlap. When we detect single-pixel overlap, shift all FOVs to the left & bottom of that position by the required distance (e.g. by ~0.1um). But I'm not sure how well this generalizes. Here are a few of the concerns:
1) Changing metadata is not really what we want to be doing
2) Search-first data will not come in a clear grid pattern, thus the patterns wouldn't be as regular (though a single pixel overlap due to the lack of stage position precision could still occur and we probably would also want to shift that).
3) At some point, we'll tackle Yokogawa data with overlap (https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/10). But thinking about this now, my suggested approach there would also be to calculate the overlap, shift the FOVs to avoid the overlap, then save the data.

---
Having reflected a bit on these concerns, maybe everything comes together as an addition to the metadata parsing.

Maybe we have an unintentional case of "imaging with overlap", but we can actually implement a general solution that would also work for that scenario.

The rough idea: 
We parse the metadata as before. But then we run another calculation on the data that will do the following: Check if any FOVs overlap, shift by however much is necessary to make the fields not overlap anymore. Save this new position as a separate column in the metadata. Any processing could then use that information (e.g. to store the data according to the new positions). Only potential stitching approaches (modality 2 here: https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/11) would require the original metadata. But given that it's still technically there, nobody can complain about not having the original metadata.

**Comment by jluethi** [comment]
tl;dr: I think we can tackle this as an additional processing during the metdata parsing

And tackling this, if done correctly, will be useful for handling images with overlap
